### ISSUE TITLE
Remove <schemaSpec> from model.divPart and add to small set of elements

### ISSUE THREAD

**Issue by jamescummings** [TEIC/TEI#1921]
schemaSpec should only be allowed in a small number of places. Currently it is allowed in:

> core: item note q quote said stage
> drama: castList epilogue performance prologue set view
> figures: cell figure
> header: change encodingDesc handNote licence scriptNote
> msdescription: accMat acquisition additions collation condition custEvent decoNote filiation foliation layout musicNotation origin provenance signatures source summary support surrogates typeNote
> namesdates: occupation
> tagdocs: specGrp
> textcrit: lem rdg
> textstructure: argument body div div1 div2 div3 div4 div5 div6 div7 epigraph postscript
> transcr: metamark

I would say that this should only be:
> encodingDesc, body, div

This enables it everywhere people actually use it, for example, in a back/div/schemaSpec but removes it from stupid places like 'provenance', 'epigraph', or dangerously 'specGrp'. This is because schemaSpec claims membership of model.divPart and it should not and just be added directly to div and body (keeping model.econdingDescPart is fine, it is needed there for some uses).

This issues specifically forces council to consider the ridiculousness of the willy-nilly application of classes through laziness.  I do *not* think that there needs to be any backwards compatibility concerns and it is a corrigible error with no deprecation needed. *If* someone comes forward and says "Oh no, my schemaSpecs inside 'metamark' are broken!", or whatever then either they should be corrected as to where to put them or council could add that in the next release. This would also be a good thing to show that the TEI is tightening the ridiculous places some elements are allowed.

**Comment by sydb** [comment]
I agree completely w/ @jamescummings except that I would add `<back>` to the list of elements of which `<schemaSpec>` can be a child. (I know, e.g., that @martindholmes usually writes his ODD files with all the prose in `<body>` and the `<schemaSpec>` in `<back>`, and I think there is a lot to be said for that approach.)

**Comment by jamescummings** [comment]
If it is allowed in `<back>` then it should also be allowed in `<front>`. 

In both cases I'd have been likely to put it in a `<div>` with the assumption that one day I might have other sections in my `<back>`. ;-)

**Comment by martindholmes** [comment]
I hadn't seen this issue till the pull request, I don't think, but if I had I would have said this requires deprecation, or at least a message to TEI-L asking whether it will invalidate a bunch of existing schemas. Since schemaSpec has no semantic import from the point of view of document encoding itself, its location is basically irrelevant, so although it seems unlikely that anyone will have put a schemaSpec inside a metamark, I can imagine scenarios where people might have sprinkled little schemaSpecs throughout a piece of structured documentation. Worth an email to the list, at least.

**Comment by sydb** [comment]
Council VF2F decides that (sigh) we will deprecate this for a short period of ½ or 1 year. To do so I plan to block #2031 and create a new branch for the deprecation itself.

Because we have no automated method for a deprecation like this, it will just be a Schematron check (as opposed to a `@validUntil`). That means that:

1. The deprecation will not appear in the appendix.
1. The build process may not remind us when it is time to end the deprecation (I forget at the moment). 
1. The check will have to be examined by hand and possibly modified every time a new element is added to the _Guidelines_.)

**Comment by sydb** [comment]
Item 3 in my previous post is idiotic. Of course the list of elements where `<schemaSpec>` should **not** be may change, but the list of elements where it is going to be allowed will most likely remain constant.

**Comment by sydb** [comment]
BTW, PR #2041 addresses this issue, and is currently awaiting review.

**Comment by lb42** [comment]
You could make exactly the same argument about many other elements. For example, the *Spec elements are members of model.inter, and can thus appear in all sorts of unlikely places, as well as some quite sensible ones.  If you really care, then surely the right solution is to define a new class for these ODD-specific components and add references to that to the more generic ones? Or just live with the fact that the TEI content models are permissive to the point of occasional embarassment.

**Comment by sydb** [comment]
Progress report — PR #2041 was merged back on 18 Feb, and is in the released version (4.2.1). Thus the next step is to undo the warning messages from that PR, and change the schema itself (per #2031) on or about 2021-10-23 (i.e., after the next release, but before the one after that).

**Comment by sydb** [comment]
The more I mess around with this, the more I think to myself that `<schemaSpec>` should just be a member of model.resourceLike, and nothing else. Sigh.

**Comment by ebeshero** [comment]
I tend to agree since that would make things a lot easier, but seems like that wouldn’t permit use in body (let alone front or back).

**Comment by sydb** [comment]
Right. All TEI customizers, including ourselves, would have to change _all_ ODD files so that `<schemaSpec>` was a direct child of `<TEI>`, and prose that goes with it was in a sibling `<text>`. But that just makes more _sense_, doesn’t it?

**Comment by martindholmes** [comment]
I also agree, as you know. There is the issue of where it would end up being output in the HTML. I'm not sure whether odd processing does different things depending on where your schemaSpec is; we should check that.

**Comment by sydb** [comment]
To be fair, I am not actually recommending we _do_ that. At least not until the dust has settled on limiting the crazy places `<schemaSpec>` can go (i.e., this ticket).
As for processing, current processing of `<schemaSpec>` to RNG and Schematron would not be affected at all, I don’t think. Might have crazy impact on the HTML documentation output, though.

**Comment by martindholmes** [comment]
I think we'd need a long period of deprecation if your intention were to prevent the use of schemaSpec elsewhere. But I think it would be a good first step to make it a member of model.resourceLike and change all our examples to encourage placing it as a direct child of TEI. As such, does order matter? If so, what order should it be in relative to the other resourceLikes?

**Comment by laurentromary** [comment]
Keeping the possibility of having it in a div would allow one to keep this editorial facility. I would also not be eager to go through all my ODDs and change everything. Life is short.

**Comment by martindholmes** [comment]
@laurentromary Definitely. This does need a long period of deprecation and lots of warning. It will affect the majority of projects, I think. I was assuming the initial move would be non-backwards-breaking.

**Comment by lb42** [comment]
This is, sorry Syd, a really daft idea. The whole point of "ODD" is that the definition of a schema and of its documentation form a single unit. Requiring schemaspec to be hived off as some other kind of resource gives entirely the wrong message.  A schemaSpec is a special kind of div -- incomplete in itself, part of a larger entity -- in my view: hence it should be permissible within body, back, or front. I am really finding it hard to see what problem this proposal is trying to solve.

**Comment by martindholmes** [comment]
@lb42 I think the issue is that schemaSpec doesn't seem like it's really part of a text, so it feels strange for it to live in the front, or the body, or the back, of anything. Prose documentation is clearly a sort of born-digital text, but the schema definition is something distinct. I'm not sure whether I agree with that or not; in processing, the two get blended in all sorts of ways, so you could argue that they're interdependent and between them they form a text of sorts, but it does seem strange to claim that a schemaSpec is some kind of front matter or back matter.

**Comment by bansp** [comment]
This ticket is an interesting read. From a petty issue of decorum to massive backwards incompatibility caused by a difficulty in tightening the encoding to be meticulous enough. It feels like a luxury of systems where there is nothing to be done but tinker with something that ain't really broken. Is the TEI such a system already? ;-) 

I've just looked at some language-resource-related standards, with stamps by OASIS and ISO, where a `schemaSpec` quite happily resides in a `div`. So now, the relevant committees, liaised with the TEI-C (one would hope that means something) are to initiate several standards-revision processes, because creating the backwards clash "would make things a lot easier" for the Council? Where's the community process in that, please? :-) 

Perhaps an alternative to the difficulties in tightening the schema to `front` and `back` is a simple "wontfix"?

**Comment by martindholmes** [comment]
I would like to have schemaSpec available as a direct child of front, body and back, but it's hardly a major inconvenience to have to put it in a div. I agree that there's little to gain from breaking backwards compatibility on such a huge scale just for a sense of tidiness.

**Comment by lb42** [comment]
That which is permitted within div (e.g. p) is also permitted with a front, body, or back devoid of divs. No need to require  a div.

**Comment by ebeshero** [comment]
I believe you misunderstood me, @bansp . I did not mean that it would be easier for Council, as it would certainly not be easier to break that much backwards compatibility. It would be easier in the sense of being a simpler model for schemaSpec from the start of things (if it had been designed that way from the beginning). We don't really have the "luxury" that you describe, no.

**Comment by lb42** [comment]
"Prose documentation is clearly a sort of born-digital text, but the schema definition is something distinct." Err, no.  Please re-read any foundational text in literate programming.

**Comment by martindholmes** [comment]
@lb42 You're right, I just tried it. But I'm sure this wasn't always the case; I can remember having to use a div to put my schemaSpec into the back element many times.

**Comment by martindholmes** [comment]
@lb42 Literate programming is a collection of good ideas which strongly influenced the way ODD was conceived, but those ideas aren't laws of nature, and the Guidelines themselves (along with most documentation people create from ODD files) has clear segmentation between the specs and the prose.

**Comment by sydb** [comment]
@martindholmes : What version of P5 are you using? The change to permit `<schemaSpec>` as a direct child of `<front>` and `<back>` was merged just over 12 hours ago. (2021-05-20T12:36Z, I think.) I think you will find it was not valid before that. I have only tested releases [1.0.0](https://www.tei-c.org/Vault/P5/1.0.0/xml/tei/custom/schema/relaxng/tei_all.rnc), [2.0.0](https://www.tei-c.org/Vault/P5/2.0.0/xml/tei/custom/schema/relaxng/tei_all.rnc), [3.0.0](https://www.tei-c.org/Vault/P5/3.0.0/xml/tei/custom/schema/relaxng/tei_all.rnc), [4.0.0](https://www.tei-c.org/Vault/P5/4.0.0/xml/tei/custom/schema/relaxng/tei_all.rnc), and [4.2.2](https://www.tei-c.org/Vault/P5/1.0.0/xml/tei/custom/schema/relaxng/tei_all.rnc), but it is not valid in `<back>` in any of those.

**Comment by sydb** [comment]
@bansp, I am not quite sure what you are talking about. I was merely musing that `<schemaSpec>` makes more _sense_ in model.resourceLike. I was explicit that I do not actually think we should _do_ that, precisely because it is backwards incompatible in a way that is 100% likely to cause users problems. (Whereas that our schemas allow `<schemaSpec>` in, say, `<handNote>` seems to me to be just a corrigible error we should be embarrassed about, and removing that laxness in the schema does not strike me as a change that is at all likely to cause anyone a problem. Not so for `<body>` or `<div>` in any way.)
But I do not think it would make anything particularly easier on Council if we were to restrict `<schemaSpec>` to model.resourceLike, and making it the case would be a significant pain for Council.
Anyway, it seems like a P6 thing to me.

**Comment by sydb** [comment]
@lb42: Well, yes, ODD is based on literate programming, and yes the “whole point of "ODD" is that the definition of a schema and of its documentation form a single unit”. But that does not mean putting `<schemaSpec>` into only model.resourceLike is daft. (That said, the fact that it violates backwards compatibility so severely may render it daft.) The prose description of, and instructions for using, an XML encoding language and the PureODD schema that declares its precise syntax are related to one another in somewhat the same way as a logical `<text>` is related to the `<sourceDocc>` or `<facsimile>` that represents the same document — different takes on the same thing.
Furthermore, a) nothing in the proposal to limit `<schemaSpec>` to only model.resourceLike would stop or even slow down a TEI customizer from using a very clear literate encoding system (after all, `<specGrpRef>` does not care if its target is in the same resource or a different one); and b) nothing in the current system prevents or even slows down a TEI customizer from violating the principles of literate encoding, and writing most of her documentation in an entirely separate document from her PureODD schema.

**Comment by martindholmes** [comment]
@sydb That explains it. Bleeding-edge plugin.:)
